File: Estrutura_de_Dados/2021-05-20/Counting_Sort/withChar/java/src/CountingSort.java
package src;

public class CountingSort {
  public static char[] order(char[] array) {
    int size = array.length;

    char[] output = new char[size];
    int[] count = new int[256];

    for (char item : array) {
      int code = (int) item;
      count[code]++;
    }

    for (int i = 1; i < 256; ++i) {
      count[i] += count[i - 1];
    }

    for (var i = size - 1; i >= 0; i--) {
      output[count[(int) array[i]] - 1] = array[i];
      count[(int) array[i]] -= 1;
    }

    return output;
  }
}

File: Estrutura_de_Dados/2021-05-20/Counting_Sort/withChar/javascript/counting-sort.js
function countingSort(array) {
  const size = array.length;

  const output = Array.from({ length: size }, (_, i) => "");
  const count = Array.from({ length: 256 }, (_, i) => 0);

  array.forEach((item) => {
    count[item.charCodeAt(0)] += 1;
  });

  for (var i = 1; i < 256; ++i) {
    count[i] += count[i - 1];
  }


  for (var i = size - 1; i >= 0; i--) {
    output[count[array[i].charCodeAt(0)] - 1] = array[i];
    count[array[i].charCodeAt(0)] -= 1;
  }

  return output;
}

module.exports.countingSort = countingSort;
